tensorpipe/common/cuda_host_allocator.cc

```diff
@@ -51,70 +55,62 @@ void CudaHostAllocator::close() {
     return;
   }
   closed_ = true;
-  cv_.notify_all();
+  processAllocations(std::move(lock));
 }
 
 void CudaHostAllocator::join() {
   close();
   if (!joined_.exchange(true)) {
-    thread_.join();
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [this]() {
+      return pendingAllocations_.empty() && (allocatedChunks_ == 0);
+    });
   }
 }
 
-void CudaHostAllocator::allocLoop(std::unique_lock<std::mutex> lock) {
-  for (;;) {
-    if (closed_ && pendingAllocations_.empty()) {
-      break;
+void CudaHostAllocator::processAllocations(std::unique_lock<std::mutex> lock) {
+  while (!pendingAllocations_.empty()) {
+    auto& callback = pendingAllocations_.front();
+    if (closed_) {
+      callback(TP_CREATE_ERROR(CudaHostAllocatorClosedError), nullptr);
```

Comment:
Shouldn't we release the lock while we invoke the callback? This has bitten us in the past...

Reply:
Got rid of the lock altogether.